Controllers branch

YarekTenko opened the pull request:
This branch implements the controller and simple views for the User model.

qr8r reviewed (changes requested):
Needs a rebase and includes commits from master

qr8r commented on config/routes.rb:
```diff
@@ -1,3 +1,5 @@
 Rails.application.routes.draw do
```
The commit message for this commit does not accurately describe the contents of the commit.

qr8r commented on config/routes.rb:
```diff
@@ -1,3 +1,5 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+  root 'users#index'
```
I think changing this route for testing is fine. But I can't really see why we would want the root of our website to be a list of users (im open to discussing this).

Likely we want to leave root as the default for now, until we create a PagesController to handle static pages like our homepage

qr8r commented on config/routes.rb:
```diff
@@ -1,5 +1,6 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
   root 'users#index'
+  get 'users/new'
```
This is unneeded. The line below defines all of your RESTful routes for your user model.

To check which routes you have defined in your app you can run `bundle exec rake routes`

qr8r commented on db/schema.rb:
```diff
@@ -12,18 +12,18 @@
 
 ActiveRecord::Schema.define(version: 20170127000826) do
 
-  create_table "outcomes", force: :cascade do |t|
```
This change is not pertinent to the feature you are implementing and probably doesnt need to be included in this commit

qr8r commented on app/views/users/new.html.erb:
```diff
@@ -0,0 +1,19 @@
+<%= form_for :user, url: users_path do |f| %>
```
You can pass an actual model to the form builder and it will figure out which url to use based on the model you pass it 

`form_for @user do |f|`

qr8r commented on app/views/users/index.html.erb:
```diff
@@ -0,0 +1,20 @@
+<% @users.each do |user| %>
+  <tr>
+    <td>
+      <%= render plain: user.username %>
```
You only need render if you are interpreting the contents of a template.

`<%= user.username %>` should be sufficient here

qr8r commented on app/views/users/index.html.erb:
```diff
@@ -0,0 +1,20 @@
+      <%= render plain: user.username %>
+    </td>
+    <td>
+    <%= render plain: user.email %>
```
^^

qr8r commented on app/views/users/index.html.erb:
```diff
@@ -0,0 +1,20 @@
+      Password:
+    </th>
+  </tr>
+<% @users.each do |user| %>
```
Excellent use of looping here! 

Be careful of your indentation. This loop and its contents should be 2 spaces indented from the `<table>` tag

qr8r commented on app/views/users/index.html.erb:
```diff
@@ -0,0 +1,20 @@
+<table>
+  <tr>
```
If you wanted to be fancy you could use the `<thead>` and `<tbody>` HTML5 tags

qr8r reviewed (changes requested):
There is no UsersController included in this PR

qr8r commented on app/views/users/new.html.erb:
```diff
@@ -0,0 +1,19 @@
+<%= form_for :user, url: @user do |f| %>
```
This can just be `<%= form_for @user do |f| %>`

qr8r commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,23 @@
+class UsersController < ApplicationController
```
There are no blank views included in this commit

qr8r commented on config/routes.rb:
```diff
@@ -1,5 +1,4 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
-  root 'users#index'
```
This change to `root` occured in the previous commit. Can you squash this removal into the previous commit?

qr8r commented on db/schema.rb:
```diff
@@ -12,11 +12,6 @@
 
 ActiveRecord::Schema.define(version: 20170127000826) do
 
-  create_table "outcomes", force: :cascade do |t|
```
Im still pretty sure you didnt mean to remove the `outcomes` table from your database schema here. You might want to kill this removal from this commit

qr8r reviewed (changes requested):
Cool! Looking really goo and this definitely works. Try and be more aware of how you are describing a commit (with the commit message) and what code you are actually putting into the commit. If a chunk of code doesn't quite fit with the contents of the commit you're working on, it's totally cool to break it into 2 commits. 

Remember to add a PR description please :)

LeonTenko committed:
Add UsersController

Added empty UsersController.

LeonTenko committed:
Define index method in UsersController

Added index method and index.html.erb view for UsersController

LeonTenko committed:
Define new method in UsersController

Add new method and new.html.erb for UsersController

LeonTenko committed:
Define create method in UsersController

Introduced user_params strong parameter guard for create method in
UsersController.

qr8r commented on config/routes.rb:
```diff
@@ -1,3 +0,0 @@
-Rails.application.routes.draw do
```
Why are you deleting the routes file?

qr8r reviewed (changes requested):
Really great job taking the feedback about smaller commits to heart. This PR is a lot easier to read now. 

Be careful not to break commits into such small chunks that the code no longer works.

Commits should be the smallest possible chunk of code *that is still functional*

for example. Adding the controller full of blank actions with out the associated views wont work. If you attempt to visit any of the pages you will get a template not found error.

Another example of over splitting code is separating the `user_params` strong params guard from the `create` action where it is actually used. In this case, at the point when the `user_params` method is added, it is never used anywhere, this is what we call "zombie code" code that exists but never runs.

Your commits are definitely way better, it just takes time to understand what constitutes the "smallest possible coherent and functional commit"
